### PR TITLE
Generation / FIX: Fix multi-device generation

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes failing tests for multi-device (e.g. Multi-GPU, GPU + CPU etc) generation. The fix is simply to make sure `pad_token_id` and all other special tokens are initialized on the correct device (e.g. for models offloaded on CPU `self.device` return `"meta"` which breaks the generation after 😢 )

cc @gante @ArthurZucker

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @gante !

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1511,7 +1518,12 @@ def generate(
         accepts_attention_mask = "attention_mask" in set(inspect.signature(self.forward).parameters.keys())
         requires_attention_mask = "encoder_outputs" not in model_kwargs
         kwargs_has_attention_mask = model_kwargs.get("attention_mask", None) is not None
-        self._prepare_special_tokens(generation_config, kwargs_has_attention_mask)
+
+        device = None
+        if "input_ids" in model_kwargs and isinstance(model_kwargs["input_ids"], torch.Tensor):
+            device = model_kwargs["input_ids"].device
+
+        self._prepare_special_tokens(generation_config, kwargs_has_attention_mask, device=device)
```
Comment (gante) [inline src/transformers/generation/utils.py:1526]: If I get it right: the `device` comes from the main model input, and not from the model itself. 

Assuming what I wrote above is correct, we should get the `device` variable after the `_prepare_model_inputs` call, which extracts the main model input from the different keywords we might see (for instance, `Whisper` does not use `input_ids`). In that case, I would move these lines to after L1532 (currently `batch_size = inputs_tensor.shape[0]`), and use `device=inputs_tensor.device` :D

**Review by gante — COMMENTED** [review]
Added a suggestion to enable this change on all modalities!

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/generation/utils.py]: Makes totally sense! Done!

**Review by gante — APPROVED** [review]
perfect, thank you for iterating 👌

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -476,6 +476,7 @@ def _prepare_attention_mask_for_generation(
         )
         can_infer_attention_mask = is_pad_token_in_inputs * is_pad_token_not_equal_to_eos_token_id
         attention_mask_from_padding = inputs.ne(pad_token_id).long()
+
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:479]: weird that this is changed 😄

**Review by ArthurZucker — APPROVED** [review]
Thanks for fixing. A small test is welcome (instead of the slow one!) to make sure we catch this earlier!

**Comment by younesbelkada** [comment]
thanks ! I don't think we can add tests as they would require a GPU, this is implictly tested through our models + quantization slow tests, hence how I catched the bug

**Comment by ArthurZucker** [comment]
ok if there is no way to repro with a minimal trick putting the weights on meta device voluntarily!

**Pull request merged**